server/src/main/java/io/crate/analyze/GeneratedColumnValidator.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.analyze;
+
+import java.util.List;
+
+import io.crate.exceptions.ColumnValidationException;
+import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.SymbolVisitor;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.Reference;
+import io.crate.metadata.RelationName;
+
+final class GeneratedColumnValidator {
+
+    private GeneratedColumnValidator() {}
+
+    private static final InnerValidator INNER_VALIDATOR = new InnerValidator();
+
+    public static void validate(Symbol symbol,
+                                RelationName relationName,
+                                String columnName,
+                                List<ColumnIdent> generatedCols) throws UnsupportedOperationException {
+        if (symbol != null) {
+            symbol.accept(INNER_VALIDATOR, new ValidatorContext(relationName, columnName, generatedCols));
+        }
```

Comment:
Could probably use `RefVisitor.visitRefs` instead of the `InnerValidator`, that would use the `DefaultTraversalSymbolVisitor` ensuring that all kinds of symbol types are traversed correctly.
Creating a ad-hoc consumer that captures the outer variables should cost about the same as creating the validator context.


Given that `symbol` is not annotated as Nullable, is the null check necessary? I think either it should have the annotation, or remove the check/replace it with an assertion that the symbol is not null.

Reply:
thx!